app/models/template.rb

```diff
@@ -73,8 +73,99 @@ def ignore_default
     self
   end
 
+  # Set attributes based on template +text and metadata found in this +text
+  # the metadata parsing can be adjusted using +options
+  def import_without_save(text, options = {})
+    self.template = text
+    @importing_metadata = self.class.parse_metadata(text)
+    Foreman::Logging.logger('app').debug "setting attributes for #{self.name} with id: #{self.id || 'N/A'}"
+    self.snippet = !!@importing_metadata[:snippet]
+
+    import_locations(options)
+    import_organizations(options)
+    import_custom_data(options)
+
+    self
+  end
+
+  # Set template attributes
+  #
+  # based on +name it either finds existing template or builds a new one
+  # then it applies changes to it and return this object, note no changes were saved at this point
+  def self.import_without_save(name, text, options = {})
+    template = self.unscoped.find_or_initialize_by(:name => name)
+    Foreman::Logging.logger('app').debug "#{template.new_record? ? 'building new' : 'updating existing'} template"
+    template.import_without_save(text, options)
+  end
+
+  # Pull out the first erb comment only - /m is for a multiline regex
+  def self.parse_metadata(text)
+    extracted = text.match(/<%\#[\t a-z0-9=:]*(.+?).-?%>/m)
+    extracted.nil? ? {} : YAML.load(extracted[1]).with_indifferent_access
+  rescue RuntimeError => e
+    Foreman::Logging.exception('invalid metadata', e)
+    {}
+  end
+
+  # Updates template metadata and save! it
+  #
+  # options can contain following keys
+  #   :force - set to true if you want to bypass locked templates
+  #   :associate - either 'new', 'always' or 'never', determines when the template should associate objects based on metadata
```

Comment:
Here is the help that needs to be in API apipie and not here.

Reply:
keeping it here too as this is method API description, it's not clear what `option` can contain as it blindly pass it elsewhere, though this is the public facing method...

anyway good point and copied the info to API definition too